### PR TITLE
Add React.memo

### PR DESCRIPTION
**Pull request by jamesgpearce**
See https://reactjs.org/docs/react-api.html#reactmemo

This function wraps a stateless functional component to memoize the result based on the props passed in.

Adds a test (and of course lots of line number noise in the others)

Addresses #7093

### CODE REVIEW

**Comment by jamesgpearce** [comment]
cc @gaearon ;)

File: lib/react.js [hunk 1]
```diff
@@ -230,8 +230,11 @@ declare module react {
   ): React$ElementFactory<ElementType>;
   declare export function createRef<ElementType: React$ElementType>(
   ): {current: null | React$ElementRef<ElementType>};
-
   declare export function isValidElement(element: any): boolean;
+  declare export function memo<Props>(
+    type: React$StatelessFunctionalComponent<Props>,
```
Comment (sebmarkbage) [inline lib/react.js]: Technically any react element type is allowed. Not sure functional components.

**Comment by jbrown215** [comment]
There are a few more fundamental issues blocking memo that I'm working on. See https://github.com/facebook/flow/issues/7093 for details. A comprehensive solution is in the works! It will work for `memo`, `lazy`, and `forwardRef`

**Comment by jamesgpearce** [comment]
You got it :)

**Pull request closed without merging**